Enrich dimension targets in click_behavior with stage-numbers

metamben opened the pull request:
Fixes #49110

### Description

Like #49256, but adding stage-numbers in dimension targets in click behavior settings.

### TODO

- [x] Make stage number depend on the target of the click, not on the dashcard's card.
- [x] Add path `["visualization_settings" "click_behavior" "parameterMapping" ["target" "dimension"]]` to the migration.
- [x] Only migrate if `click_behavior.linkType = "question"`.

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

metamben committed:
Enrich dimension targets in click_behavior with stage-numbers

Fixes #49110

piranha commented:
Do you think it needs to be a migration rather than an augment during read?.. I mean data migrations are finicky enough to potentially be a problem, and we already have `migrate-viz-settings` for stuff like that. I see that there is another migration merged in, but it's also not released yet, so it's more or less in the same condition. 😁

kamilmielnik commented on src/metabase/db/custom_migrations.clj:
```diff
@@ -1683,6 +1683,17 @@
       (when (not= pmappings orig-pmappings)
         pmappings))))
 
+(defn- card->last-stage
```
:x:  `stage-number` is `0` for multi-stage queries

1. `git checkout master`
2. `yarn dev`
3. https://github.com/user-attachments/assets/973ffbce-3d9a-4eee-9a00-22d7db075026
4. kill `yarn dev`
5. `git checkout 49110-click-behavior-stage-number-migration`
6. `yarn dev`
7. https://github.com/user-attachments/assets/bf937294-f280-4d5f-bc2c-5893e02b889f

kamilmielnik replied:
This is fixed now

metamben commented:
> Do you think it needs to be a migration rather than an augment during read?.. I mean data migrations are finicky enough to potentially be a problem, and we already have `migrate-viz-settings` for stuff like that. I see that there is another migration merged in, but it's also not released yet, so it's more or less in the same condition. 😁

Yes, we had an extended discussion about this and the outcome was that not doing this would be more error prone.

metamben committed:
Address TODOs

- Make stage number depend on the target of the click, not on the dashcard's card.
- Add path ["visualization_settings" "click_behavior" "parameterMapping" ["target" "dimension"]] to the migration.
- Only migrate if click_behavior.linkType = "question".

metamben committed:
Merge branch 'dashboard-filter-columns' into 49110-click-behavior-stage-number-migration

kamilmielnik commented on src/metabase/db/custom_migrations.clj:
```diff
@@ -1683,6 +1683,17 @@
         pmappings))))
 
+(defn- card->last-stage
+  [record]
```
:x: Parameter mapping disappears in the UI after running the migration.
Reason unclear.
It may have something to do with object keys not matching `target.dimension` - we should regenerate those.

![image](https://github.com/user-attachments/assets/e011d46b-9668-47d1-882f-6d82fe00805f)

metamben replied:
Added conversion for the `id` fields and keys.

kamilmielnik replied:
This is fixed now

metamben committed:
Rewrite IDs and keys in parameterMappings too

metamben committed:
Comment the test

kamilmielnik reviewed (approved):
I did some testing and didn’t encounter any issues :+1:
I haven’t reviewed the code though.

metamben committed:
Merge branch 'dashboard-filter-columns' into 49110-click-behavior-stage-number-migration

bshepherdson commented on src/metabase/db/custom_migrations.clj:
```diff
@@ -1743,3 +1747,131 @@
+
+(defn- update-viz-settings-target-dimensions
+  [viz-settings click-behavior->update-fn]
+  (cond-> viz-settings
+    (-> viz-settings :column_settings map?)
+    (update :column_settings
+            update-vals
+            #(update-column-settings-target-dimensions % click-behavior->update-fn))))
```
Can this be `m/update-existing` instead?

metamben replied:
I prefer making sure here that the value is indeed a map. I don't want to mess with anything else.

bshepherdson commented on test/metabase/db/custom_migrations_test.clj:
```diff
@@ -2346,3 +2369,143 @@
+              ;; crazy as it sounds, the FE uses the JSON encoded form of the dimensions as IDs and keys
+              (let [dimension-strs (mapv json/generate-string dimensions)
+                    dimension-keys (mapv keyword dimension-strs)]
+                (fn [[target-id0 target-id1 target-id2]]
```
This inner function is really hard to follow, mostly because of the vertical length. Can we either abstract an inner chunk into a separate function (perhaps from index to `:parameterMapping` map?), turn it into a `for` or similar, or use the horizontal space a bit more effectively?

metamben committed:
Merge branch 'dashboard-filter-columns' into 49110-click-behavior-stage-number-migration

metamben committed:
Address review feedback: use less vertical space

metamben committed:
Merge branch 'dashboard-filter-columns' into 49110-click-behavior-stage-number-migration
